Refactored bool TypeChecker::visit(FunctionCall const& _functionCall).

svenski123 opened the pull request:
Logic is now split into the following methods:
- TypeChecker::typeCheckABIDecodeAndRetrieveReturnType (unchanged)
- TypeChecker::typeCheckTypeConversionAndRetrieveReturnType (added)
- TypeChecker::typeCheckFunctionLikeAndRetrieveReturnType (added)
- TypeChecker::typeCheckNonABIDecodeFunctionLike (added)

<!--### Your checklist for this pull request

Please review the [guidelines for contributing](http://solidity.readthedocs.io/en/latest/contributing.html) to this repository.

Please also note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.
-->

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### Description
FunctionCall AST nodes can represent type conversion, struct construction and function call expressions, each of which require different type check and annotation logic and which was handled in a single large and unwieldy method.

Logic common to all FunctionCall nodes such as visiting descendent AST nodes, evaluating argument types for positional funciton expressions, assignment of annotations and initial determinination of FunctionCall sub-type (i.e. type conversion, function call, etc.) was left in TypeChecker::visit(FunctionCall const&).

Error checking and return type determination logic was moved to distinct methods and for function call/struct ctors return type determiniation was separated from the rather more lengthy error checking logic.

Function calls of kind ABIDecode were already delegated to a separate method TypeChecker::typeCheckABIDecodeAndRetrieveReturnType(...) which was left unchanged.
 <!--
Please explain the changes you made here.

Thank you for your help!
-->

svenski123 committed:
Refactor of bool TypeChecker::visit(FunctionCall const& _functionCall).

Visit method now cleanly determines if node represents a function call,
struct construction or type conversion. Type checking, validation and
error message logic is moved to separate methods.

chriseth commented:
Thanks for your pull request! Your version of `TypeChecker::visit(FunctionCall` now looks very clean! Reviewing this properly will take some time, though, I hope you don't mind. Also note that you have a build error:
```
/root/project/libsolidity/analysis/TypeChecker.cpp: In member function 'dev::solidity::TypePointer dev::solidity::TypeChecker::typeCheckFunctionLikeAndRetrieveReturnType(const dev::solidity::FunctionCall&, dev::solidity::FunctionTypePointer)':
/root/project/libsolidity/analysis/TypeChecker.cpp:1815:18: error: unused variable 'allowDynamicTypes' [-Werror=unused-variable]
   if (bool const allowDynamicTypes = m_evmVersion.supportsReturndata())
                  ^~~~~~~~~~~~~~~~~
```

svenski123 commented:
Sorry – my build box is Centos7 w/stock GCC 4.8.5; it built cleanly on that.

I’ve reproduced w/devtoolset-7 (GCC 7.3.1) and pushed fix to github.

Note I’ve also checked ‘test/soltest -- --no-ipc’ passes (this is with GCC 4.8.5 build).

Please do take all the time you need to review – I understand.

 

Oddly I did see GCC 4.8.5 fail on a similar error in boost 1.66 for which I patched the boost header locally

i.e. boost/optional/optional.hpp:954:5: error: unused parameter ‘rhs’ [-Werror=unused-parameter]

It’s not clear to me why GCC 4.8.5 then didn’t complain about TypeChecker.cpp.

From: chriseth [mailto:notifications@github.com] 
Sent: 19 October 2018 10:15
To: ethereum/solidity
Cc: svenski123; Author
Subject: Re: [ethereum/solidity] Refactored bool TypeChecker::visit(FunctionCall const& _functionCall). (#5275)

 

Thanks for your pull request! Your version of TypeChecker::visit(FunctionCall now looks very clean! Reviewing this properly will take some time, though, I hope you don't mind. Also note that you have a build error:

/root/project/libsolidity/analysis/TypeChecker.cpp: In member function 'dev::solidity::TypePointer dev::solidity::TypeChecker::typeCheckFunctionLikeAndRetrieveReturnType(const dev::solidity::FunctionCall&, dev::solidity::FunctionTypePointer)':
/root/project/libsolidity/analysis/TypeChecker.cpp:1815:18: error: unused variable 'allowDynamicTypes' [-Werror=unused-variable]
   if (bool const allowDynamicTypes = m_evmVersion.supportsReturndata())
                  ^~~~~~~~~~~~~~~~~

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/ethereum/solidity/pull/5275#issuecomment-431299160> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AJyYsAah8KWFJOAOZR_jasgO41a7yUsIks5umZglgaJpZM4XvWP5> .  <https://github.com/notifications/beacon/AJyYsNR43vNKixGHNXcysaXpK5Cy6Gjxks5umZglgaJpZM4XvWP5.gif>

chriseth commented on libsolidity/analysis/TypeChecker.h:
```diff
@@ -94,7 +94,28 @@ class TypeChecker: private ASTConstVisitor
+		bool _abiEncoderV2
+	);
+
+	/// Performs type checks and determine result types for type conversion FunctionCall nodes
```
Should be `determines`

chriseth commented on libsolidity/analysis/TypeChecker.h:
```diff
@@ -94,7 +94,28 @@ class TypeChecker: private ASTConstVisitor
+		TypePointer const _expressionActualType
+	);
+
+	/// Performs type checks and determine result types for function call & struct ctor FunctionCall nodes
```
Should be `determines` and please don't use `&` as abbreviation.

chriseth commented on libsolidity/analysis/TypeChecker.h:
```diff
@@ -94,7 +94,28 @@ class TypeChecker: private ASTConstVisitor
+		FunctionTypePointer _functionType
+	);
+
+	/// Performs type checks on function call & struct ctor FunctionCall nodes not of kind ABIDecode.
```
Should be `determines` and please don't use `&` as abbreviation.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1947,86 +2014,174 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
+	// For positional calls only, store argument types
+	if (_functionCall.names().empty())
+	{
+		shared_ptr<TypePointers> const argumentTypes = make_shared<TypePointers>();
```
Can you move from it properly if it is const?

svenski123 replied:
Good catch, it's actually the shared_ptr copy ctor that gets called rather than the move ctor which is why this compiled cleanly.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1947,86 +2014,174 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
+
+		if (actualType->category() == Type::Category::Struct)
+		{
+			FunctionTypePointer const functionType =
```
I think this would still make sense as a function-level variable. This would also allow you to perform a generic assignmont to `isPure` for all cases at the end.

svenski123 replied:
Done - visit() is rather less "branchy" now.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1947,86 +2014,174 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
+				dynamic_cast<StructType const&>(*actualType).constructorType();
+			_functionCall.annotation().isPure = argumentsArePure;
+			_functionCall.annotation().kind = FunctionCallKind::StructConstructorCall;
+			_functionCall.annotation().type = typeCheckFunctionLikeAndRetrieveReturnType(
```
Shouldn't "struct constructor" be its own thing?

svenski123 replied:
I'd say it's debatable, at type checking level constructors and functions are syntically very similar - and the logic applied to constructors is more similar to vanilla functions than some of the special function types (i.e. ABI decode, etc.).

In general, my preference is for less code and more data driven validation - i.e. validating number and types of arguments against parameter definitions in the function type object rather than handling each function type kind specifically; for example, how the four BareCall and friends are treated to append extra text to the message.

Of course custom error messages may be considered necessary despite the increased ad-hoc logic - as this is not for me judge, I purposefully tried to ensure that all generated error messages were unchanged.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,161 +1681,198 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-	/// For error message: Struct members that were removed during conversion to memory.
-	set<string> membersRemovedForStructConstructor;
-	if (_functionCall.annotation().kind == FunctionCallKind::StructConstructorCall)
+	if (_functionType->kind() == FunctionType::Kind::ABIDecode)
```
I think it would be better to pull this out into the calling function

svenski123 replied:
I've moved the ABIDecode case handling up to visit().

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,161 +1681,198 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-		m_errorReporter.typeError(_functionCall.location(), "Type is not callable");
-		_functionCall.annotation().type = make_shared<TupleType>();
-		return false;
+		TypeChecker::typeCheckNonABIDecodeFunctionLike(_functionCall, _functionType);
```
I think it would be better to give this a more generic name - "NonABIDecodingFunctionLike" does not sound like it would cover 99% of the cases :)

svenski123 replied:
I have renamed the method.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1890,52 +1923,86 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		{
 			auto const& argType = type(*arguments[i]);
-			if (functionType->takesArbitraryParameters() && i >= parameterTypes.size())
+			if (_functionType->takesArbitraryParameters() && i >= parameterTypes.size())
```
I just re-checked: The only functions where `takesArbitraryParameters()` returns true are the encoding and decoding functions. Perhaps it would make sense to factor them out into a single checking function?

svenski123 replied:
Argument type/count checking is now in a standalone method which handles additional arguments (i.e. arbitrary arguments beyond the function specified parameters). As the arbitrary parameters are not used in any other checking, I don't think it makes sense to separate out encoding functions for that reason alone.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,161 +1681,198 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+		);
 
-	if (!functionType->padArguments())
+	if (!_functionType->padArguments())
```
This is only true for the `abi.encodePacked` function.

svenski123 replied:
The if condition would be true for all eight of the following FunctionType::Kind values; it's not apparent to me that the other seven values would be filtered out in advance at this point.

    case Kind::BareCall:
    case Kind::BareCallCode:
    case Kind::BareDelegateCall:
    case Kind::BareStaticCall:
    case Kind::SHA256:
    case Kind::RIPEMD160:
    case Kind::KECCAK256:
    case Kind::ABIEncodePacked:

chriseth replied:
You are right, I forgot about the other 7. The thing is, though, that all except `ABIEncodePacked` only have a single parameter of type `bytes memory` (we changed that recently) and rational number literals cannot be converted to `bytes memory`. So the check below actually only is meaningful for `ABIEncodePacked`.

svenski123 replied:
I've moved variadic and packed argument checks to a separate method for the various ABI encode* functions.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1877,10 +1907,13 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 	}
-	else if (!functionType->takesArbitraryParameters() && parameterTypes.size() != arguments.size())
+
+	if (!_functionType->takesArbitraryParameters() && parameterTypes.size() != arguments.size())
```
Starting from here, the only thing that happens is checking that the number and names of arguments is correct and that their types match, rigth? Perhaps this would be a nice candidate for extracting into a function whose names clearly states everything it does.

svenski123 replied:
I have moved argument/parameter checking into a separate method TypeChecker::typeCheckFunctionCallCompatibleArguments().

chriseth reviewed (changes requested):
I added some comments but did not check the changes in every detail yet.

svenski123 commented:
I've pushed a new commit addressing the issues raised in the review.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2132,55 +2118,93 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
+		actualType = typeType->actualType();
+
+	// Determine annotation kind for this FunctionCall node
+	const FunctionCallKind annoKind =
```
The fact that this will be stored in the AST annotations later is secondary, I would say. So what about `functionCallKind` or `callKind` as name?

svenski123 replied:
Done.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2132,55 +2118,93 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
+			nullptr;
+
+	// Determine purity; for function calls, this also depends upon the callee and its FunctionType
+	const bool annoPure = functionType && annoKind == FunctionCallKind::FunctionCall
```
Do you really need this intermediate variable?

svenski123 replied:
No; removed.

svenski123 commented:
I've pushed a new commit addressing comments raised.

svenski123 commented:
@chriseth I'd like to close this PR and bounty out - could you kindly review and advise what else may be needed?

svenski123 commented:
Appveyor check is now fixed.
There was one review comment which I may have missed - I have now responded to it.
Please kindly review and respond.

chriseth commented:
Please rebase and remove the merge commit - it makes it impossible to review properly.

svenski123 commented:
I've removed the merge commit, rebased to head of develop and force pushed the branch back to gitbub.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+
+	// Determine type for type conversion or struct construction expressions
+	if (expressionType->category() == Type::Category::TypeType)
+		if (auto const* const typeType = dynamic_cast<TypeType const*>(expressionType.get()))
```
Should not be conditional but rather throw if conversion fails:
```suggestion
		actualType = dynamic_cast<TypeType const&>(*expressionType)).actualType();
```

svenski123 replied:
This was set early on and kept for type conversions to be passed in later on (null for vanilla functions).
I've moved it down into the block that handles type conversions/struct ctors - it will throw there if null.

The separate type check method for type conversions now extracts the expression type from the function call node itself rather than relying on the visit function to pass it in - I believe this is cleaner.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+	// Determine function call kind for this FunctionCall node
+	const FunctionCallKind functionCallKind =
+		!actualType
+			? FunctionCallKind::FunctionCall
```
We usually but separators (`?` and `:`) at the end of the line.

svenski123 replied:
Understood.
This code has been replaced with an if statement.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+	TypePointer actualType;
+
+	// Determine type for type conversion or struct construction expressions
+	if (expressionType->category() == Type::Category::TypeType)
```
This ``if`` is redundant due to the ``dynamic_cast`` in the next line.

svenski123 replied:
Checking category() first avoids the overhead of dynamic casts on all function call nodes.
I would normally use a static cast instead of dynamic here, as it appears category() == TypeType if and only if expressionType is pointing to a TypeType instance.

However as dynamic casts are used throughout TypeChecker I left it as a dynamic cast.
Happy to make it a static cast or take out the category() check - please advise.

chriseth replied:
The choice is between dynamic cast of pointer or check plus dynamic cast of reference. Since dynamic cast of pointer is shorter, I would prefer that.

svenski123 replied:
The choices I was suggesting were either dynamic cast of pointer or check plus static cast - the second being in this case:

    if (expressionType->category() == Type::Category::TypeType)
        actualType = static_cast<TypeType const*>(expressionType.get())->actualType();

chriseth replied:
Readability and safety go before efficiency. If you want to verify the correct relation between the runtime type and the result of `category()` then please use a condition on `category()` plus a dynamic cast of reference type (which throws if the argument's type does not match).

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+
+	// Determine function call kind for this FunctionCall node
+	const FunctionCallKind functionCallKind =
+		!actualType
```
IMHO this is actually *less* readable than before - I'd suggest avoiding large and nested ternary expressions and moving this logic back to the ``if`` above (or some ``if``'s here).

I think continuing to determine kind and type together here will in fact make things easier.

svenski123 replied:
The reason for the terniary operands was to allow making the kind/type variables const and assigning them once (and in one place) in their intialisers, as I found the original code rather hard to follow.

Agreed however that it should be more readable - I've rewritten as shown below which reads a bit more clearly.

(Also unrelated realised my comment above regarding dynamic cast versus category() + static cast applies here as well.)

    // Determine function call kind and function type for this FunctionCall node
    FunctionCallKind functionCallKind = FunctionCallKind::Unset;
    FunctionTypePointer functionType = nullptr;

    if (!actualType)
    {
        functionCallKind = FunctionCallKind::FunctionCall;
        functionType = dynamic_pointer_cast<FunctionType const>(expressionType);
    }
    else if (actualType->category() == Type::Category::Struct)
    {
        functionCallKind = FunctionCallKind::StructConstructorCall;
        functionType = dynamic_cast<StructType const&>(*actualType).constructorType();
    }
    else
    {
        functionCallKind = FunctionCallKind::TypeConversion;
        functionType = nullptr;
    }

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-		_functionCall.annotation().type = make_shared<TupleType>(returnTypes);
+	{
+		annoType = make_shared<TupleType>();
+		m_errorReporter.typeError(_functionCall.location(), "Type is not callable");
```
Could you move this closer to the definition of `functionType`? In the current form I would wonder why it is OK to assign `nullptr` to `functionType`.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+	// Determine function type for function call or struct construction expressions
+	FunctionTypePointer const functionType =
+		functionCallKind == FunctionCallKind::StructConstructorCall
+			? dynamic_cast<StructType const&>(*actualType).constructorType()
```
also here

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+
+	// Determine purity; for function calls, this also depends upon the callee and its FunctionType
+	_functionCall.annotation().isPure =
+		functionType && functionCallKind == FunctionCallKind::FunctionCall
```
This might be easier to read with an actual `if`, but I guess this is a matter of stye. In any case, please move the separators to the end.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+		: argumentsArePure;
+
+	// Update annotations in FunctionCall node (needed by typeCheck...() methods called below)
+	_functionCall.annotation().kind = functionCallKind;
```
Why not do that as soon as `functionCallKind` is determined?

svenski123 replied:
Initially when I untangled the visit method I isolated determination of function call kind, function type, purity, return type and assignment to the node's annoation into separate steps. Some of this been folded back togther and this now looks out of place.

The last commit has removed the annotation temporaries (functionCallKind, etc.) so this is gone.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-			_functionCall.annotation().kind = FunctionCallKind::TypeConversion;
-
-	}
+	TypePointer resultType = move(_expressionActualType);
```
```suggestion
	TypePointer resultType = _expressionActualType;
```
(it is const)

svenski123 replied:
TypePointer itself is a non-const shared pointer to a const Type i.e. 'std::shared_ptr<Type const>' and not 'std::shared_ptr<Type const> const'. Moving it as opposed to copying it means the shared_ptr's reference count block does not need to be dereferenced, locked and incremented.

However in the latest commit, I've removed the _expressionActualType parameter from typeCheckTypeConversionAndRetrieveReturnType() which now extracts the expression result type itself. This was to remove the function level actualType variable from the the visit method.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-bool TypeChecker::visit(FunctionCall const& _functionCall)
+TypePointer TypeChecker::typeCheckTypeConversionAndRetrieveReturnType(
+	FunctionCall const& _functionCall,
+	TypePointer const _expressionActualType
```
```suggestion
	TypePointer const _targetType
```
("actual type" is not the right name in this context, I would argue)

chriseth replied:
Or you can also just call it `_resultType` and make it non-const.

svenski123 replied:
As mentioned in the comment above, _expressionActualType parameter is now gone, typeCheckTypeConversionAndRetrieveReturnType() extracts the expression result type itself now and it is assigned to 'resultType' which is certainly a better name than 'actual type'.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+	_functionCall.annotation().kind = functionCallKind;
+
+	// Determine return types
+	TypePointer annoType;
```
Maybe this should in fact be named ``returnType``?

svenski123 replied:
Done.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+	for (size_t i = 0; i < param_arg_map_size; ++i)
+		if (!type(*param_arg_map[i])->isImplicitlyConvertibleTo(*parameterTypes[i]))
+		{
+			solAssert(!!param_arg_map[i], "unmapped parameter");
```
Isn't this too late for the assertion? ``param_arg_map[i]`` was already dereferenced in the ``if`` condition...

svenski123 replied:
Yes it is - fixed.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+{
+	solAssert(!!_functionType, "");
+	solAssert(
+		_functionType->kind() != FunctionType::Kind::ABIEncode ||
```
These should be equalities, not inequalities, right? This way the assertion is always true and checks nothing.

svenski123 replied:
Yes good catch!

ekpyron reviewed (commented):
I think this looks rather solid and is a huge improvement already - I left some minor comments, but I'll need another run at it to be absolutely sure that it's consistent with the old version.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-				resultType = make_shared<AddressType>(payable ? StateMutability::Payable : StateMutability::NonPayable);
+				solAssert(dynamic_cast<ContractType const*>(resultType.get())->isPayable(), "");
+				solAssert(
+					dynamic_cast<AddressType const*>(argType.get())->
```
Could you split at `<` instead of at `->`, or perhaps just once at the `,`?

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+					auto const* identifier =
+						dynamic_cast<Identifier const*>(arguments.front().get())
+				)
+					if (
```
The line break here is not really necessary.

svenski123 replied:
The CODING_STYLE.md said lines should be no more than 99 characters and without the break it came to 102 hence I broke the line.

I'll keep it on one line and in future allow for a few characters over 99 before breaking.

chriseth replied:
Ah weird, I thought the line would be shorter without the line break. Choose whatever you think is better!

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+		switch (functionType->kind())
+		{
+		case FunctionType::Kind::ABIDecode:
+			{
```
We usually do not indent inside a case.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
 				_functionCall.location(),
-				"Named arguments cannot be used for functions that take arbitrary parameters."
+				parameterNames.size() > argumentNames.size()
+				? "Some argument names are missing."
```
Again: separators at the end

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-	}
+
+	// Parameter to argument map
+	size_t const param_arg_map_size = parameterTypes.size();
```
Please use camelCase, or someone could also argue that this variable is not necessary at all (don't mind the complexity).

svenski123 replied:
It used to be a plain C array (i.e. no size()) hence the size variable.
Removed the size var and camel cased param_arg_map.

chriseth reviewed (changes requested):
Wonderful! I also only have some small remarks.

svenski123 commented:
@ekpyron Thank you for your review, I've just checked in commit addressing them all.
@chriseth THank you for the additional comments, I'm reviewing them now and will check in one further commit once I'm done.

svenski123 commented:
Hi - @ekpyron @chriseth, thank you both again for your reviews.

I have commited the last revision of this for tonight which I believe addresses all issues that have been raised. I have worked the visit() method a bit more and the logic flow is now quite clean IMHO.

I have left comments marked REVIEW regarding handling of non-callable nodes (i.e. errors) - in the original code, such function call nodes would have been marked as FunctionCallKind::FunctionCall but the functionType would have been null; an empty type tuple return type would have been assigned and "Type is not callable" would have been reported to the error reporter.

In the new logic, such nodes are marked as FunctionCallKind::Unset - however to avoid triggering the assert I force the kind back to FunctionCallKind::FunctionCall to maintain the previous behaviour. Clearly attempting to call a non-callable expression should be a normal compile time warning, not a failed assertion - but I don't know if leaving the node's function call kind set to Unset would cause issues elsewhere.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2230,23 +2230,24 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		// REVIEW - original code set function call kind to function call for non-callables
-		_functionCall.annotation().kind = FunctionCallKind::FunctionCall;
-	}
+		funcCallAnno.kind = FunctionCallKind::FunctionCall;
```
I agree, it is probably better to set this to `FunctionCall`, but in this case, please add a `solAssert(m_errorReporter.hasError(), "")` to ensure that we already reported an error earlier.

svenski123 replied:
Thanks, I have committed that change.

ekpyron reviewed (dismissed):
I looked over it again and haven't found any more issues!

chriseth commented:
Nice! Is it possible to squash some of the commits (especially the "review feedback" ones)?

svenski123 commented:
Great, I have squashed the PR branch into one commit, rebased it on to the head of develop and forced it back to github. This should now hopefully be ready to go/merge.
